Rewrite the character introduction

aphillips opened the pull request:
- switch to typographic unit in place of user-perceived character
- change example to use a Balinese word
- add illustrations
- remove extraneous languages from grapheme example

Fixes #150, Addresses i18n-action/#160


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/aphillips/bp-i18n-specdev/pull/153.html" title="Last updated on Mar 13, 2025, 7:11 PM UTC (44a96b4)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/bp-i18n-specdev/153/d268561...aphillips:44a96b4.html" title="Last updated on Mar 13, 2025, 7:11 PM UTC (44a96b4)">Diff</a>

aphillips committed:
Replace Bangla example, replace user-perceived character

- switch to typographic unit in place of user-perceived character
- change example to use a Balinese word
- add illustrations
- remove extraneous languages from grapheme example

Fixes #150, Addresses i18n-action/#160

r12a commented on index.html:
```diff
@@ -1275,29 +1275,39 @@ <h2>Characters</h2>
+    <p>At their simplest, typographic units are a single shape that can be tied one-to-one to the underlying computing representation. But a typographic unit can be formed, in some scripts, from more than one character. And a given logical character can take many different shapes due to such influences as font selection, style, or the surrounding context (such as adjacent characters). In some cas ...
 
-    <p>When user-perceived characters are represented visibly (on screen or in print), they are represented by individual rendering units. This visual unit is called a [=grapheme=] (the word [=glyph=] is also used). Graphemes are the visual units found in fonts and rendering software.</p>
+    <p>When typographic units are represented visibly (on screen or in print), they are represented by individual rendering units. This visual unit is called a [=grapheme=] (the word [=glyph=] is also used). Graphemes are the visual units found in fonts and rendering software.</p>
```
> This visual unit is called a [=grapheme=] (the word [=glyph=] is also used). Graphemes are the visual units found in fonts and rendering software.

I think this is incorrect.  'Grapheme' is a (rather fuzzy) word used by linguists and such to talk about text, which may be written using a pen or quill, not just in type, and that approximates to a user-perceived character.  The glossary definition describes it as something "that a typical _user_ would perceive as being a single unit" (my emphasis).  The individual rendering units are just glyphs.

The specdev text is not consistent, either, because after example 6 it goes on to equate 'grapheme' with 'grapheme clusters', rather than 'visual units found in fonts'.

I think that the change were trying to get away from the notion of user-perceived characters, but the text seems to have simply substituted the term 'grapheme' for the former.  If we're going to do this thoroughly, we may need to define grapheme clusters in terms of typographic units, rather than graphemes. The correlation between the two is already made further up.  We should probably also make that change in the glossary definition for grapheme cluster, too, though.

aphillips replied:
Thanks for this review.

You're right. This is a mess... so in the latest revision I more-or-less rewrote the section entirely. I need to go back and make sense of other sections too.

r12a commented on index.html:
```diff
@@ -1275,29 +1275,39 @@ <h2>Characters</h2>
+
+	<aside class="example" title="Difference between Unicode grapheme cluster and a typographic unit">
+	<p>In Balinese, the word "ᬅᬲ᭄ᬢ", which means "to be", is composed of this character sequence: 
+	<span class="codepoint" translate="no"><bdi lang="ban">&#x1B05;</bdi><code class="uname">U+1B05 BALINESE LETTER AKARA</code></span>
```
I think this would be better as a (vertical) list, rather than inline.

aphillips replied:
done

r12a commented on index.html:
```diff
@@ -1275,29 +1275,39 @@ <h2>Characters</h2>
+	<span class="codepoint" translate="no"><bdi lang="ban">&#x1B32;</bdi><code class="uname">U+1B05 BALINESE LETTER SA</code></span>
+	<span class="codepoint" translate="no"><bdi lang="ban">&#x1B44;</bdi><code class="uname">U+1B05 BALINESE ADEG ADEG</code></span>
+	<span class="codepoint" translate="no"><bdi lang="ban">&#x1B22;</bdi><code class="uname">U+1B05 BALINESE LETTER TA</code></span>.
+	Currently, the rules in [[UAX29]] break this into three extended default grapheme clusters, like this:</p>
```
Drop 'default', or say 'break this by default into'

aphillips replied:
done

r12a commented on index.html:
```diff
@@ -1275,29 +1275,39 @@ <h2>Characters</h2>
+	Currently, the rules in [[UAX29]] break this into three extended default grapheme clusters, like this:</p>
+	<img src="images/balinese-gc.png">
+		
+	<p>However, the <em>adeg adeg</em> (which functions similar to a [=virama=]) should result in the just two typographic units, like this:</p>
```
the just -> just just

aphillips replied:
done

r12a commented on index.html:
```diff
@@ -1275,29 +1275,39 @@ <h2>Characters</h2>
+	<img src="images/balinese-ortho-syllab.png">
+	</aside>
+		
+	<p>For more information, see our article <a href="https://www.w3.org/International/articles/definitions-characters/index.en.html#characters">Character encodings: Essential concepts</a>.</p>
```
This has now been updated to reflect the new 15.1 rules.

We also have another article that is relevant here: [Typographic character units in complex scripts](https://www.w3.org/International/questions/qa-indic-graphemes)

aphillips replied:
thanks! added the link. put both in the link (purple) box at the top

r12a commented on index.html:
```diff
@@ -1275,29 +1275,39 @@ <h2>Characters</h2>
+		
+	<p>For more information, see our article <a href="https://www.w3.org/International/articles/definitions-characters/index.en.html#characters">Character encodings: Essential concepts</a>.</p>
 
     <p>Graphemes are encoded into computer systems using "logical characters". A <a>character set</a> is a set of logical characters: a specific collection of characters that can be used together to encode text. The most important character set is the <a>Universal Character Set</a>, also known as [[Unicode]]. This character set includes all of the characters used to encode text, including histori ...
```
What's a 'logical' character?

aphillips replied:
I made all-gone with this text

aphillips committed:
A thorough and comprehensive rewrite

- Incorped @r12a's suggestions
- Added links
- Fixed the 'bigtext' style to be less space hoggy
- Rewrote basically everything

aphillips committed:
Fix proofreading errors

aphillips committed:
More typos, plus rewrite the code unit/encoding material

aphillips committed:
Fix one errant mention of 'grapheme cluster' ahead of the definition

aphillips committed:
Fix other instances of the word 'grapheme'

Some of these changes are substantive

aphillips committed:
Tweak formatting on one advisement in truncation section

aphillips committed:
Add files to echidna, update truncation

Fixes #49

aphillips committed:
Further improve the truncation examples.

- Make the emoji family example into an aside with an ID for easy
  reference
- Split the long example of truncation into two examples
- Replace the byte truncation example with the "I can eat glass" version

aphillips committed:
Fix missing link.

aphillips committed:
Fix broken link

aphillips committed:
First pass: replace typo unit with visual text unit

- Also updates the family emoji example

aphillips committed:
Develop "I can eat glass" example of truncation

aphillips committed:
middle of *the* child

aphillips committed:
Copy edit

aphillips committed:
Rewrite the string truncation section

- Add proporptional font image
- Remove the truncation image
- Rewrite all of the examples
- Add a best practice not to use visual units

aphillips committed:
Tidy example 17

aphillips committed:
Further tighten example 17

aphillips commented:
@r12a I would appreciate a fresh review here. The #characters section now uses _visual text unit_, but the bigger news is in #char_truncation, which I did a significant rewrite of. I changed the examples, including removing and replacing the new one I just added that we took from webauthn. If you have the time, I'd appreciate comment and to see what you think of the changes.

aphillips committed:
Say 'legacy character encoding'

aphillips committed:
Further improve the Tamil examples

aphillips committed:
Fix typo ('over' => 'other')

aphillips committed:
Merge charmod-based mustard with char_def

r12a commented on index.html:
```diff
@@ -1261,66 +1261,68 @@ <h2>Characters</h2>
+<aside class="links" id="links_char_def">
 <p class="links_title">Useful background and overviews for this section</p>
 <ul>
 <li class="w3"><p class="link"><a href="https://www.w3.org/TR/charmod/">Character Model for the World Wide Web: Fundamentals</a> — basic guidelines related to the use of characters and encodings.</p></li>
```
Perhaps we should add a sentence after "... and encodings." to say that most of the content has been ported here, and so no need to go there ?

aphillips replied:
done

r12a commented on index.html:
```diff
@@ -1261,66 +1261,68 @@ <h2>Characters</h2>
-		</p>
-		<p>The last example, in the Devanagari script (which is used to write Hindi, among other languages) is made up of four graphemes:</p>
-		<p class=bigtext lang=hi>&#x092F;&#x0942;&nbsp;&#x0928;&#x093F;&nbsp;&#x0915;&#x094B;&nbsp;&#x0921;</p>
+    <p>A <dfn class="lint-ignore">visual text unit</dfn> is a single visible "character" of text. This can be on a screen or in other media, such as printed on paper or written on the back of a napkin. Such units are a matter of perception and the term is necessarily imprecise, particularly when it comes to writing systems that use a combination of combining marks, complex positioning, or complex ...
```
How about: "a single visible "character" of text." -> "a single unit of the visible text as perceived by a user."

r12a commented on index.html:
```diff
@@ -1261,66 +1261,68 @@ <h2>Characters</h2>
-		<p class=bigtext lang=hi>&#x092F;&#x0942;&nbsp;&#x0928;&#x093F;&nbsp;&#x0915;&#x094B;&nbsp;&#x0921;</p>
+    <p>A <dfn class="lint-ignore">visual text unit</dfn> is a single visible "character" of text. This can be on a screen or in other media, such as printed on paper or written on the back of a napkin. Such units are a matter of perception and the term is necessarily imprecise, particularly when it comes to writing systems that use a combination of combining marks, complex positioning, or complex ...
+    
+    <p>There are some other terms with similar meanings. The term [=grapheme=] is sometimes used by linguists (and others) to refer to what a typical user would perceive as a single unit. The term [=glyph=] is used by typographers (and others) to refer to an individual rendering unit. Sometimes a [=visual text unit=] is said to be a <dfn class="lint-ignore">user-perceived character</dfn>, althoug ...
```
```suggestion
    <p>... as some users in certain scripts will "perceive" (read) a particular shape in the text as separate "characters", such as when certain South Asian scripts use "stacks".</p>
```

I think what you might be trying to say (?) here is something like:

> ... for example some users of Brahmi-derived scripts will perceive (read) components in a stacked consonant cluster as separate visual text units, whereas others will see the whole stack as one visual text unit.

aphillips replied:
done

r12a commented on index.html:
```diff
@@ -1261,66 +1261,68 @@ <h2>Characters</h2>
+		<p><span class="bigtext" lang="hi">&#x092F;&#x0942;&nbsp;<span style="color:red">&#x0928;&#x093F;</span>&nbsp;<span style="color:black">&#x0915;&#x094B;</span>&nbsp;<span style="color:magenta">&#x0921;</span></p>
     </aside>
+	
+    <p>[=Visual text units=] are the visual representation of an underlying character sequence encoded in the computer's memory. In order to encode characters for use by software, we have to start by agreeing on what is being encoded.</p>
```
I think that the first sentence essentially repeats what is said in the para before the example, except that it doesn't mention that it could represent a single character.

aphillips replied:
The redundancy was on purpose, but I removed it and rephrased

r12a commented on index.html:
```diff
@@ -1261,66 +1261,68 @@ <h2>Characters</h2>
-		<p>Returning to the example above, the Hindi word for Unicode is made of four graphemes:</p>
-		<p class=bigtext lang=hi>&#x092F;&#x0942;&nbsp;<span style="color:red">&#x0928;&#x093F;</span>&nbsp;&#x0915;&#x094B;&nbsp;&#x0921;</p>
-		<p>Several of these graphemes are made up of more than one Unicode character because of the way that the Devanagari script works. In Devanagari, the basic set of "letters" are syllables ending with the short 'a' vowel sound. When you want to use a different vowel, you add a combining vowel character that changes the shape of the grapheme. The red text in the example above is the syllable "ni" i ...
+    <p>The relationship between [=code points=] and [=visual text units=] can be complex. In most cases, a sequence of [=code points=] that forms a single [=visual text unit=] should be processed as a single, indivisible textual unit. For example, when cursoring across text, the cursor should "jump across" or select the entire visual text unit (and its underlying code points) together. It shouldn ...
```
I think you should provide a  counter example, such as forward vs backward deletion, to underline the point you are making that the definition can vary according to the editing operation.

aphillips replied:
I tried to address this, but it's awfully hard to make this clear.

r12a commented on index.html:
```diff
@@ -1261,66 +1261,68 @@ <h2>Characters</h2>
+	
+    <aside class="example" title="Example of the difference between grapheme clusters and code points">
+		<p>Returning to [[[#visual-text-unit-example]]] above, the Hindi word for "Unicode" is made of four visual text units:</p>
+		<p class="bigtext" lang="hi">&#x092F;&#x0942;&nbsp;<span style="color:red"><strong>&#x0928;&#x093F;</strong></span>&nbsp;&#x0915;&#x094B;&nbsp;&#x0921;</p>
```
I suspect we'll have an accessibility issue here. How will a user with a screen reader see which characters you are highlighting?  I wonder whether using the `<mark>` element would work better?  We can style that however we want (although a background colour change may be easier on the eye than the multi-colour and bolding effects?).

r12a replied:
> U+0928 (the syllable "na") and U+093F 

Why not use `<span class="codepoint" translate="no"><bdi lang="ml">&#x0928;</bdi> <span class="uname">U+0928 DEVANAGARI LETTER NA</span></span> (the syllable "na") and <span class="codepoint" translate="no"><bdi lang="ml">&#x093F;</bdi> <span class="uname">U+093F DEVANAGARI VOWEL SIGN I​</span></span>`

r12a commented on index.html:
```diff
@@ -1334,22 +1336,81 @@ <h2>Characters</h2>
+	<p>Currently, the rules in [[UAX29]] break this by default into three grapheme clusters, like this:</p>
+	<img src="images/balinese-gc.png">
+		
+	<p>However, the <em>adeg adeg</em> (which functions similar to a [=virama=]) should result in just two visual text units, like this:</p>
```
> the <em>adeg adeg</em> (which functions similar to a [=virama=]) should result in

maybe

> the use of <em>adeg adeg</em> to create a stack (which functions similar to a [=virama=]) should result in

r12a commented on index.html:
```diff
@@ -1334,22 +1336,81 @@ <h2>Characters</h2>
+        </table>
+        
+        
+        <p>Many common emoji can <em>only</em> be formed using sequences of code points, but should be treated as a single [=visual text unit=] when displaying or processing the text. You wouldn't want to put a line-break in the middle of the family!</p>
```
The emoji produced in these 4 examples is inconsistent, and the last example has text flowing over the graphics.

aphillips replied:
Your browser must render this differently. On FF/Windows and Chrome/Windows it works fine. My Linux FF does fail here (I suppose I need to install the color emoji font).

After last week's call, I dithered about replacing this example with graphics and that's probably what I should do.

![image](https://github.com/user-attachments/assets/28c3db5f-486b-4e2b-910a-460908a06181)

r12a replied:
Here's what i see (on all browsers on my Mac).  I recommend that you use images.

<img width="750" alt="Screenshot 2025-03-13 at 14 14 22" src="https://github.com/user-attachments/assets/2e26eb05-d3b7-4383-bf2b-3dcb321ec1b4" />

aphillips replied:
Yes. That's what I got on Linux.

r12a commented on index.html:
```diff
@@ -1334,22 +1336,81 @@ <h2>Characters</h2>
+    
+    <p>Unicode [=code points=] are not the values actually used when processing or serializing text in a computing system. Text processing instead depends on sequences of fixed-size integer units called [=code units=]. There are different sizes of [=code unit=]. The most familiar of these is called a <dfn class="lint-ignore">byte</dfn> or <em>octet</em>, consisting of 8 bits.</p>
+    
+    <p>There are different sized [=code units=] used by different runtime environments. Other common sizes include 16- or 32-bit units. On the Web, the most familiar of these are the 16-bit units used by the [[DOM]], JavaScript, and the various string types in [[INFRA]]. These use 16-bit code units processed according to the rules of the UTF-16 [=character encoding=] of [[Unicode]]. (See [[[#char ...
```
'There are...' alert!

Suggestion:

> Different sized [code units](https://www.w3.org/TR/i18n-glossary/#dfn-code-unit) are used by different runtime environments.

r12a commented on index.html:
```diff
@@ -1334,22 +1336,81 @@ <h2>Characters</h2>
+    
+    <p>There are different sized [=code units=] used by different runtime environments. Other common sizes include 16- or 32-bit units. On the Web, the most familiar of these are the 16-bit units used by the [[DOM]], JavaScript, and the various string types in [[INFRA]]. These use 16-bit code units processed according to the rules of the UTF-16 [=character encoding=] of [[Unicode]]. (See [[[#char ...
+    
+    <p>A [=character encoding form=] (or [=character encoding=] for short) are the rules for converting between [=code points=] in a [=character set=] and the [=code units=] used to store and process text. Non-Unicode character encoding forms are called [=legacy character encodings=]. For information on choosing a character encoding, see: [[[#char_choosing]]].</p>
```
a character encoding form *is* the set of rules

or a character encoding form establishes the set of rules

r12a commented on index.html:
```diff
@@ -1261,66 +1261,68 @@ <h2>Characters</h2>
 <div id="char_checklist"></div>
 </details-->
 
+<section id="char_def" class="subtopic">
```
You seem to have lost the 'See also' panel that is in the original chapter text.

aphillips replied:
We have a link to that section in the body of the text. I could make see-also of that link and the one on choosing encodings in appropriate places instead of in-lining the reference.

r12a commented:
We seem to be missing the mustard: 
> Specifications SHOULD use specific terms, when available, instead of the general term 'character'.
Is that intentional?

r12a commented:
This PR is extremely long and complicated, and appears to be growing more so.  Why don't we merge it after incorporating the changes suggested here, and make any further alterations on a new PR?  We can then compare the /TR version, the ED version, and the PR to spot changes more easily.

r12a commented on index.html:
```diff
@@ -1381,61 +1442,31 @@ <h2>Characters</h2>
 	</details>
 	</div>
 
-<p>Computer storage and communication rely on units of physical storage and information interchange, such as bits and bytes (8-bit units, also called octets). A frequent error in specifications and implementations is the equating of characters with units of physical storage. The mapping between characters and such units of storage is actually quite complex.</p>
```
Not sure why this was dropped. It seems useful.

aphillips replied:
I like this text, but... we just had several paragraphs about code points and code units. This might be moved to be an intro to that bit (just after the emoji composition example where it starts "Unicode code points are not the values...")?

r12a commented:
And adding the truncation text to the already VERY long PR for the character definition section is not making you many friends in my neck of the woods ;-). Not to mention the increased possibility for conflicts to arise if anyone else is working on the text.

r12a commented:
I may make detailed review of the truncation section later, but in the meantime here are some thoughts.

Truncation may be applied because (a) there's a limit to the size of stored strings, (b) there's a limit to what is visible on the UI.  These different issues are mentioned in the text, but don't (from a rapid scan) appear to be clearly distinguished and addressed.

Note to self: In addition to the impact of code unit limits on text size (example 17), there is the potential issue of text expansion (compare "I can eat glass, it does not hurt me" with "Is urrainn dhomh glainne ithe, chan eil e gam ghoirteachadh" in Scots gaelic), which may push useful bits of information out of the viewing window; and the issue of sentence syntax, which may have the same effect in languages like Hindi or Japanese, which move the verb to the end of the sentence.  Do we mention those?

aphillips commented:
> We seem to be missing the mustard:
> 
> > Specifications SHOULD use specific terms, when available, instead of the general term 'character'.
> > Is that intentional?

I think it is redundant with this one:

> Specifications SHOULD use the term [code point](https://www.w3.org/TR/i18n-glossary/#dfn-code-point) instead of the term 'character'. If the term 'character' is used, it MUST be explicitly defined to mean a Unicode [code point](https://www.w3.org/TR/i18n-glossary/#dfn-code-point). The term [Unicode Scalar Value](https://www.w3.org/TR/i18n-glossary/#dfn-scalar-value) MAY also be used.

aphillips committed:
Revert truncation section

aphillips committed:
Address comments

aphillips committed:
Final tweaks

aphillips committed:
Replace family emoji examples with images

aphillips commented:
Per the 2025-03-13 call, I'm going to merge this to the editors copy.
